modules/auxiliary/server/capture/ldap.rb

```diff
@@ -0,0 +1,271 @@
+                suitable_response(pdu.app_tag)
+              )
+            end
+
+      on_send_response(client, res) unless res.nil?
+    rescue StandardError => e
+      client.close
+      print_error("Failed to handle LDAP request due to #{e}")
+    end
+  end
+
+  def report_cred(opts)
+    service_data = {
+      address: opts[:ip],
+      port: opts[:port],
+      service_name: opts[:service_name],
+      protocol: 'tcp',
+      workspace_id: myworkspace_id
+    }
+
+    credential_data = {
+      origin_type: :service,
+      module_fullname: fullname,
+      username: opts[:user],
+      private_data: opts[:password],
+      private_type: :password
+    }.merge(service_data)
+
+    if opts[:domain]
+      credential_data = {
+        realm_key: Metasploit::Model::Realm::Key::ACTIVE_DIRECTORY_DOMAIN,
+        realm_value: opts[:domain]
+      }.merge(credential_data)
+    end
+
+    login_data = {
+      core: create_credential(credential_data),
+      status: Metasploit::Model::Login::Status::UNTRIED
+    }.merge(service_data)
+
+    create_credential_login(login_data)
+  end
+
+  def search_res(filter, msgid, attrflt = :all)
+    if @ldif.nil? || @ldif.empty?
+      attrs = []
+      if attrflt.is_a?(Array)
+        attrflt.each do |at|
+          attrval = [Rex::Text.rand_text_alphanumeric(10)].map(&:to_ber).to_ber_set
+          attrs << [at.to_ber, attrval].to_ber_sequence
+        end
+        dn = "dc=#{Rex::Text.rand_text_alphanumeric(10)},dc=#{Rex::Text.rand_text_alpha(4)}"
+        appseq = [
+          dn.to_ber,
+          attrs.to_ber_sequence
+        ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+        [msgid.to_ber, appseq].to_ber_sequence
+      end
+    else
+      ldif.map do |bind_dn, entry|
+        next unless filter.match(entry)
+
+        attrs = []
+        entry.each do |k, v|
+          if attrflt == :all || attrflt.include?(k.downcase)
+            attrvals = v.map(&:to_ber).to_ber_set
+            attrs << [k.to_ber, attrvals].to_ber_sequence
+          end
+        end
+        appseq = [
+          bind_dn.to_ber,
+          attrs.to_ber_sequence
+        ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+        [msgid.to_ber, appseq].to_ber_sequence
+      end.compact.join
+    end
+  end
+
+  def suitable_response(request)
+    responses = {
+      Net::LDAP::PDU::BindRequest => Net::LDAP::PDU::BindResult,
+      Net::LDAP::PDU::SearchRequest => Net::LDAP::PDU::SearchResult,
+      Net::LDAP::PDU::ModifyRequest => Net::LDAP::PDU::ModifyResponse,
+      Net::LDAP::PDU::AddRequest => Net::LDAP::PDU::AddResponse,
+      Net::LDAP::PDU::DeleteRequest => Net::LDAP::PDU::DeleteResponse,
+      Net::LDAP::PDU::ModifyRDNRequest => Net::LDAP::PDU::ModifyRDNResponse,
+      Net::LDAP::PDU::CompareRequest => Net::LDAP::PDU::CompareResponse,
+      Net::LDAP::PDU::ExtendedRequest => Net::LDAP::PDU::ExtendedResponse
+    }
+
+    responses[request]
+  end
```

Comment:
This method needs a default case or consumers of the method need to check for `nil` and respond accordingly.  As is the if the request type was not in the hash as a supported type `nil` would be returned and `encode_ldap_response` would end up passing `to_ber_appsequence(nil)` which may result in `in `+': nil can't be coerced into Integer (TypeError)`.

Reply:
For now this has been accounted for by testing `nil` on the returned result and closing the connection.  In a future update this should be adjusted to match the [RFC](https://datatracker.ietf.org/doc/html/rfc4511#section-4.1.1) and return a `Notice of Disconnection` before closing the connection:
```
If the server receives an LDAPMessage from the client in which the
   LDAPMessage SEQUENCE tag cannot be recognized, the messageID cannot
   be parsed, the tag of the protocolOp is not recognized as a request,
   or the encoding structures or lengths of data fields are found to be
   incorrect, then the server SHOULD return the Notice of Disconnection
   described in Section 4.4.1, with the resultCode set to protocolError,
   and MUST immediately terminate the LDAP session as described in
   Section 5.3.
```